contrib/apelscripts/condor_batch_blah.py

```diff
@@ -0,0 +1,170 @@
+import configparser
+from pathlib import Path
+
+
+CLI = argparse.ArgumentParser(
+    description="Generate APEL accounting records for an HTCondor CE and LRMS",
+    epilog="""
+HTCondor configuration values:
+
+  condor_config_val:
+    PER_JOB_HISTORY_DIR  path to which per-job history files are written
+
+  condor_ce_config_val:
+    APEL_OUTPUT_DIR      path to which APEL record files should be written
+    APEL_SCALE_DEFAULT   default scale when no job attribute applies, such as 1.0 or UNDEFINED
+    APEL_CE_HOST         hostname of the CE
+    APEL_CE_ID           APEL identifier for the CE
+    APEL_SCALING_ATTR    job attribute for optional performance factor
+    APEL_SPEC_ATTR       job attribute for optional absolute performance
+""".strip(),
+    formatter_class=argparse.RawDescriptionHelpFormatter,
+)
+CLI.add_argument(
+    "--apel-config",
+    help="path to apel client configuration file [default: %(default)s]",
+    default="/etc/apel/client.cfg",
+    type=Path,
+)
+CLI.add_argument(
+    "--dry-run",
+    help="do not perform destructive actions, write data to stdout",
+    action="store_true",
+)
+
+
+def read_lrms_config_val(key: str) -> str:
+    return check_output(["condor_config_val", key], universal_newlines=True).strip()
+
+
+def read_ce_config_val(key: str) -> str:
+    return check_output(["condor_ce_config_val", key], universal_newlines=True).strip()
+
+
+def read_apel_specs(apel_config: Path, ce_id: str) -> "dict[str, float]":
+    """Read APEL manual_specX entries for a given CE identifier"""
+    config = configparser.ConfigParser()
+    with apel_config.open() as config_stream:
+        config.read_file(config_stream)
+    specs = {}
+    for spec_id in range(1, len(config["spec_updater"]) + 1):
+        try:
+            spec = config["spec_updater"][f"manual_spec{spec_id}"]
+        except KeyError:
+            break
+        spec_ce_id, spec_type, spec_value, *_ = spec.split(",")
+        if ce_id == spec_ce_id:
+            specs[spec_type] = float(spec_value)
+    return specs
+
+
+def format_apel_scaling(apel_config: Path, ce_id: str) -> str:
+    """Build a ClassAd expression for the factor of used vs average performance"""
+    specs = read_apel_specs(apel_config, ce_id)
+    try:
+        scale_query = read_ce_config_val("APEL_SCALE_DEFAULT")
+    except CalledProcessError:
+        # assume average performance by default
+        scale_query = "1.0"
+    try:
+        scaling_attr = read_ce_config_val("APEL_SCALING_ATTR")
+    except CalledProcessError:
+        pass
+    else:
+        scale_query = f"({scaling_attr} ?: {scale_query})"
+    try:
+        spec_attr = read_ce_config_val("APEL_SPEC_ATTR")
+    except CalledProcessError:
+        return scale_query
+    else:
+        spec_scale_query = scale_query
+        for spec_type, spec_value in reversed(list(specs.items())):
+            spec_scale_query = (
+                f"(({spec_attr}.'{spec_type}' / {spec_value}) ?: {spec_scale_query})"
```

Comment:
Maybe I'm missing something here, but this seems like it will have the net effect of setting `spec_scale_query` to the first (and only the first) item listed in each CE's spec entry.  If that's the desired behavior, why loop over the whole list?

Reply:
The code otherwise looks fine, although I also can't tell why the RPM build fails.

Reply:
The RPM build fails on one platform, because we are trying to build for an old unsupported platform. If the review is approved, I'll use my admin privileges to merge it. I'll modify the tests to remove the obsolete ones.

Reply:
The expression is intended to go through the different spec types of the CE and use the first one which was defined on the resource as well: the `{spec_attr}.'{spec_type}'` part is undefined if the resource does not define the spec type.

Let's say the expression built is
```
((MachineAttrApelSpecs0.'HEPSPEC' / 13.45) ?: ((MachineAttrApelSpecs0.'SI2K' / 2614.0) ?: (MachineAttrApelScaling0 ?: 1.0)))
```
and the machine only defines `ApelSpecs = [SI2K=2793]`. Then the entire `(MachineAttrApelSpecs0.'HEPSPEC' / 13.45)` part is `undefined` and Elvis picks the part checking `SI2K` instead.

I'll add a comment into the code to explain this.

Reply:
Sorry, I misread the code.  I didn't understand that it was extending `spec_scale_query` by one clause each time; I thought for some reason it was replacing `spec_scale_query` on each iteration.  The changes look good.